File: day_02/puzzle_test.exs
Code.load_file("puzzle.exs", __DIR__)

ExUnit.start()

defmodule PuzzleTest do
  use ExUnit.Case

  describe "part one" do
    test "calculates checksum from input" do
      assert Puzzle.calc_part_1([
               "abcdef",
               "bababc",
               "abbcde",
               "abcccd",
               "aabcdd",
               "abcdee",
               "ababab"
             ]) == 12
    end

    test "calculates checksum from input file" do
      assert Puzzle.load_input() |> Puzzle.calc_part_1() == 6696
    end
  end

  describe "part two" do
    test "finds the common id from input" do
      assert Puzzle.calc_part_2(["abcde", "fghij", "klmno", "pqrst", "fguij", "axcye", "wvxyz"]) ==
               {{"fgij", 3}, ["fguij", "fghij"]}
    end

    test "finds the common id from input file" do
      assert Puzzle.load_input() |> Puzzle.calc_part_2() ==
               {{"bvnfawcnyoeyudzrpgslimtkj", 21},
                ["bvnfawcnyoeyudzrpgsldimtkj", "bvnfawcnyoeyudzrpgsleimtkj"]}
    end
  end
end

File: 2019/day_12/puzzle_test.rb
require 'minitest/autorun'
require_relative 'puzzle'

class PuzzleTest < Minitest::Test
  def test_calc_part_1
    # calculates examples
    assert_equal(
      179, 
      Puzzle.new(["<x=-1, y=0, z=2>","<x=2, y=-10, z=-7>","<x=4, y=-8, z=8>","<x=3, y=5, z=-1>"])
            .calc_part_1(10).last
    )

    assert_equal(
      1940,
      Puzzle.new(["<x=-8, y=-10, z=0>","<x=5, y=5, z=10>","<x=2, y=-7, z=3>","<x=9, y=-8, z=-3>"])
            .calc_part_1(100).last
    )

    # calculates from input file
    assert_equal 12466, Puzzle.new.calc_part_1(1000).last
  end

  def test_calc_part_2
    # calculates examples
    assert_equal(
      2772,
      Puzzle.new(["<x=-1, y=0, z=2>","<x=2, y=-10, z=-7>","<x=4, y=-8, z=8>","<x=3, y=5, z=-1>"])
            .calc_part_2
    )

    assert_equal(
      4686774924,
      Puzzle.new(["<x=-8, y=-10, z=0>","<x=5, y=5, z=10>","<x=2, y=-7, z=3>","<x=9, y=-8, z=-3>"])
            .calc_part_2
    )

    # calculates from input file
    assert_equal 360689156787864, Puzzle.new.calc_part_2
  end
end